homeassistant/components/mobile_app/webhook.py

```diff
@@ -0,0 +1,174 @@
+
+    webhook_payload = req_data.get(ATTR_WEBHOOK_DATA, {})
+
+    if req_data[ATTR_WEBHOOK_ENCRYPTED]:
+        enc_data = req_data[ATTR_WEBHOOK_ENCRYPTED_DATA]
+        webhook_payload = _decrypt_payload(device[CONF_SECRET], enc_data)
+
+    if (webhook_type not in WEBHOOK_TYPES and
+            device[ATTR_APP_ID] in INTEGRATIONS_MAP):
+        # Unknown webhook type, check if there's a platform
+        platform_name = INTEGRATIONS_MAP[device[ATTR_APP_ID]]
+
+        platform = get_platform(hass, DOMAIN, platform_name)
+        return await platform.async_handle_webhook_message(hass, device,
+                                                           webhook_type,
+                                                           webhook_payload)
+
+    try:
+        data = WEBHOOK_SCHEMAS[webhook_type](webhook_payload)
+    except vol.Invalid as ex:
+        err = vol.humanize.humanize_error(webhook_payload, ex)
+        _LOGGER.error('Received invalid webhook payload: %s', err)
+        return empty_okay_response(headers=headers)
+
+    if webhook_type == WEBHOOK_TYPE_CALL_SERVICE:
+        try:
+            await hass.services.async_call(data[ATTR_DOMAIN],
+                                           data[ATTR_SERVICE],
+                                           data[ATTR_SERVICE_DATA],
+                                           blocking=True,
+                                           context=device_context(device))
+        except (vol.Invalid, ServiceNotFound):
+            raise HTTPBadRequest()
+
+        return empty_okay_response(headers=headers)
+
+    if webhook_type == WEBHOOK_TYPE_FIRE_EVENT:
+        event_type = data[ATTR_EVENT_TYPE]
+        hass.bus.async_fire(event_type, data[ATTR_EVENT_DATA],
+                            EventOrigin.remote,
+                            context=device_context(device))
+        return empty_okay_response(headers=headers)
+
+    if webhook_type == WEBHOOK_TYPE_RENDER_TEMPLATE:
+        try:
+            tpl = template.Template(data[ATTR_TEMPLATE], hass)
+            rendered = tpl.async_render(data.get(ATTR_TEMPLATE_VARIABLES))
+            return json_response({"rendered": rendered}, headers=headers)
+        except (ValueError, TemplateError) as ex:
+            return json_response(({"error": ex}), status=HTTP_BAD_REQUEST,
+                                 headers=headers)
+
+    if webhook_type == WEBHOOK_TYPE_UPDATE_LOCATION:
+        await hass.services.async_call(DT_DOMAIN,
+                                       DT_SEE, data,
+                                       blocking=True,
+                                       context=device_context(device))
+        return empty_okay_response(headers=headers)
+
+    if webhook_type == WEBHOOK_TYPE_UPDATE_REGISTRATION:
+        data[ATTR_APP_ID] = device[ATTR_APP_ID]
+        data[ATTR_APP_NAME] = device[ATTR_APP_NAME]
+        data[ATTR_SUPPORTS_ENCRYPTION] = device[ATTR_SUPPORTS_ENCRYPTION]
+        data[CONF_SECRET] = device[CONF_SECRET]
```

Comment:
We cannot take in a secret from a webhook, that means that people can disable encryption if they are able to deliver a message.